### PR TITLE
Unique letters

### PR DESCRIPTION
**Pull request by phep45**

### CODE REVIEW

File: src/main/java/com/luxoft/wordscounter/EntryPair.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
+package com.luxoft.wordscounter;
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/EntryPair.java]: Use Pair from apache commons {http://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/tuple/Pair.html}

File: src/main/java/com/luxoft/wordscounter/OnPoint.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    public static void main(String... args) {
+        AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext(WordsConfig.class);
+
+        OnPoint onPoint = context.getBean(OnPoint.class);
+        onPoint.process("src\\main\\resources\\words\\test2.txt");
+
+        context.close();
+    }
+
+    public WordsCollector getWordsCollector() {
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/OnPoint.java]: You can remove it, for spring beans you do not need in this methods

File: src/main/java/com/luxoft/wordscounter/ReportPrinter.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import org.springframework.stereotype.Component;
+
+import java.util.Map;
+
+@Component
+public class ReportPrinter {
+    public void printReport(Map<SetOfPairs, Integer> map) {
+        Preconditions.checkNotNull(map, "No data to print");
+        for (Map.Entry<SetOfPairs, Integer> s : map.entrySet()) {
+            System.out.println(s.getValue() + " => " + s.getKey());
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/ReportPrinter.java]: Extract constant

File: src/main/java/com/luxoft/wordscounter/SetOfPairs.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.*;
+
+public class SetOfPairs implements Comparable<SetOfPairs> {
+    private static final Logger log = LoggerFactory.getLogger(SetOfPairs.class);
+    private Set<EntryPair> entryPairSet = new TreeSet<>();
+
+    public void insert(String letter) {
+        for (EntryPair oldPair : entryPairSet) {
+            if (oldPair.getLetter().equals(letter)) {
+                log.trace("Pair {} exists; incrementing amount", oldPair);
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/SetOfPairs.java]: good

File: src/main/java/com/luxoft/wordscounter/SetOfPairs.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
+package com.luxoft.wordscounter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+public class SetOfPairs implements Comparable<SetOfPairs> {
+    private static final Logger log = LoggerFactory.getLogger(SetOfPairs.class);
+    private Set<EntryPair> entryPairSet = new TreeSet<>();
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/SetOfPairs.java]: You can set coomporator in constractor. In this case you do not need in this class.

File: src/main/java/com/luxoft/wordscounter/SetOfPairs.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    public void insert(String letter) {
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/SetOfPairs.java]: Move this logic to serves class or below

File: src/main/java/com/luxoft/wordscounter/TextFileReader.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+
+@Component
+public class TextFileReader {
+
+    private static final String TEXT_FILE_EXTENSION = ".txt";
+    private static final String SPACE = " ";
+
+    public List<String> readFromFile(File file) throws IOException {
+        checkArgument(file.getName().endsWith(TEXT_FILE_EXTENSION), "Only TXT files supported");
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/TextFileReader.java:24]: good

File: src/main/java/com/luxoft/wordscounter/TextFileReader.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+@Component
+public class TextFileReader {
+
+    private static final String TEXT_FILE_EXTENSION = ".txt";
+    private static final String SPACE = " ";
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/TextFileReader.java]: StringUtils.SPASE {https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html}

File: src/main/java/com/luxoft/wordscounter/UniqueLetters.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+
+@Component
+public class UniqueLetters {
+
+    private static final Logger log = LoggerFactory.getLogger(UniqueLetters.class);
+    private static final String EMPTY_STR = "";
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/UniqueLetters.java]: StringUtils.EMPTY

File: src/main/java/com/luxoft/wordscounter/UniqueLetters.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    private Map<SetOfPairs, Integer> map;
+
+    public Map<SetOfPairs, Integer> count(List<String> wordsList) {
+        Preconditions.checkNotNull(wordsList);
+        map = new TreeMap<>();
+        wordsList.forEach(s -> {
+            log.trace("Processing word: {}", s);
+            List<String> letters = Arrays.asList(s.split(EMPTY_STR));
+            SetOfPairs set = new SetOfPairs();
+            letters.forEach(set::insert);
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/UniqueLetters.java]: Move insert in this class as private method

File: src/main/java/com/luxoft/wordscounter/UniqueLetters.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+    private static final Logger log = LoggerFactory.getLogger(UniqueLetters.class);
+    private static final String EMPTY_STR = "";
+
+    private Map<SetOfPairs, Integer> map;
+
+    public Map<SetOfPairs, Integer> count(List<String> wordsList) {
+        Preconditions.checkNotNull(wordsList);
+        map = new TreeMap<>();
+        wordsList.forEach(s -> {
+            log.trace("Processing word: {}", s);
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/UniqueLetters.java]: good

File: src/main/java/com/luxoft/wordscounter/UniqueLetters.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+    private static final Logger log = LoggerFactory.getLogger(UniqueLetters.class);
+    private static final String EMPTY_STR = "";
+
+    private Map<SetOfPairs, Integer> map;
+
+    public Map<SetOfPairs, Integer> count(List<String> wordsList) {
+        Preconditions.checkNotNull(wordsList);
+        map = new TreeMap<>();
+        wordsList.forEach(s -> {
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/UniqueLetters.java]: Try to use wordsList.stream()
    .collect(Collectors.toMap(....

File: src/main/java/com/luxoft/wordscounter/UniqueLetters.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+            log.trace("Processing word: {}", s);
+            List<String> letters = Arrays.asList(s.split(EMPTY_STR));
+            SetOfPairs set = new SetOfPairs();
+            letters.forEach(set::insert);
+            populateMap(set);
+        });
+        return map;
+    }
+    private void populateMap(SetOfPairs set) {
+        if (!map.containsKey(set)) {
```
Comment (ARybochenko) [inline src/main/java/com/luxoft/wordscounter/UniqueLetters.java]: reverse this logic, negative statement

File: src/main/resources/app-config.xml [hunk 14]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ARybochenko) [inline src/main/resources/app-config.xml]: Please remove it

File: src/main/resources/logback.xml [hunk 15]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ARybochenko) [inline src/main/resources/logback.xml:1]: good

File: src/test/java/com/luxoft/wordscounter/ReportPrinterTest.java [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+import org.springframework.test.context.support.AnnotationConfigContextLoader;
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration(classes =WordsConfig.class, loader = AnnotationConfigContextLoader.class)
+public class ReportPrinterTest {
+
+    @Autowired
+    ReportPrinter reportPrinter;
```
Comment (ARybochenko) [inline src/test/java/com/luxoft/wordscounter/ReportPrinterTest.java]: private

File: src/main/java/com/luxoft/wordscounter/EntryPair.java (hunk 1 above)
Comment (phep45) [inline src/main/java/com/luxoft/wordscounter/EntryPair.java]: done

File: src/test/java/com/luxoft/wordscounter/ReportPrinterTest.java [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+import org.springframework.test.context.support.AnnotationConfigContextLoader;
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration(classes =WordsConfig.class, loader = AnnotationConfigContextLoader.class)
+public class ReportPrinterTest {
```
Comment (ARybochenko) [inline src/test/java/com/luxoft/wordscounter/ReportPrinterTest.java]: I think it is a not correct class name

File: src/test/java/com/luxoft/wordscounter/ReportPrinterTest.java [hunk 18]
```diff
@@ -0,0 +1,23 @@
+package com.luxoft.wordscounter;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+import org.springframework.test.context.support.AnnotationConfigContextLoader;
+
+@RunWith(SpringJUnit4ClassRunner.class)
```
Comment (ARybochenko) [inline src/test/java/com/luxoft/wordscounter/ReportPrinterTest.java]: You do not need in spring context in this place

File: src/test/java/com/luxoft/wordscounter/UniqueLettersTest.java [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+import org.springframework.test.context.support.AnnotationConfigContextLoader;
+
+import java.util.*;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(SpringJUnit4ClassRunner.class)
```
Comment (ARybochenko) [inline src/test/java/com/luxoft/wordscounter/UniqueLettersTest.java]: Remove

File: src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+import static junit.framework.Assert.assertEquals;
+import static junit.framework.Assert.assertNotNull;
+
+@RunWith(SpringJUnit4ClassRunner.class)
```
Comment (ARybochenko) [inline src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java]: Remove

File: src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+@ContextConfiguration(classes =WordsConfig.class, loader = AnnotationConfigContextLoader.class)
+public class TextFileReaderTest {
+
+    @Autowired
```
Comment (ARybochenko) [inline src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java]: Remove

File: src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+    private TextFileReader textFileReader;
```
Comment (ARybochenko) [inline src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java:23]: Create by contractor

File: src/test/java/com/luxoft/wordscounter/UniqueLettersTest.java [hunk 23]
```diff
@@ -0,0 +1,111 @@
...
+        Map<SetOfPairs, Integer> result = uniqueLetters.count(listOfWords);
+        assertEquals(expectedMap, result);
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldThrowException() {
+        uniqueLetters.count(null);
+    }
+
+    private Map<SetOfPairs, Integer> generateExpectedMap() {
```
Comment (ARybochenko) [inline src/test/java/com/luxoft/wordscounter/UniqueLettersTest.java]: User immutable map builder

Comment (phep45) [inline src/test/java/com/luxoft/wordscounter/UniqueLettersTest.java]: done

File: src/main/resources/app-config.xml (hunk 14 above)
Comment (phep45) [inline src/main/resources/app-config.xml]: done

File: src/main/java/com/luxoft/wordscounter/UniqueLetters.java (hunk 13 above)
Comment (phep45) [inline src/main/java/com/luxoft/wordscounter/UniqueLetters.java]: done

File: src/main/java/com/luxoft/wordscounter/UniqueLetters.java (hunk 9 above)
Comment (phep45) [inline src/main/java/com/luxoft/wordscounter/UniqueLetters.java]: done

File: src/main/java/com/luxoft/wordscounter/TextFileReader.java (hunk 8 above)
Comment (phep45) [inline src/main/java/com/luxoft/wordscounter/TextFileReader.java]: done

File: src/main/java/com/luxoft/wordscounter/ReportPrinter.java (hunk 3 above)
Comment (phep45) [inline src/main/java/com/luxoft/wordscounter/ReportPrinter.java]: done

File: src/test/java/com/luxoft/wordscounter/ReportPrinterTest.java (hunk 18 above)
Comment (phep45) [inline src/test/java/com/luxoft/wordscounter/ReportPrinterTest.java]: done

File: src/test/java/com/luxoft/wordscounter/UniqueLettersTest.java (hunk 19 above)
Comment (phep45) [inline src/test/java/com/luxoft/wordscounter/UniqueLettersTest.java]: done

File: src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java (hunk 20 above)
Comment (phep45) [inline src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java]: done

File: src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java (hunk 21 above)
Comment (phep45) [inline src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java]: done

File: src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java (hunk 22 above)
Comment (phep45) [inline src/test/java/com/luxoft/wordscounter/TextFileReaderTest.java:23]: done

File: src/main/java/com/luxoft/wordscounter/OnPoint.java (hunk 2 above)
Comment (phep45) [inline src/main/java/com/luxoft/wordscounter/OnPoint.java]: done

**Pull request merged**